File: AtlasWeb/Core/Modules/Merkezi_Yonetim_Sistemi/Stok_Modulleri/Sarf_Edilebilen_Malzeme_Tanimlari/ConsumableMaterialDefinitionFormViewModel.ts
//$FA4C1124
import { BaseViewModel } from "NebulaClient/Model/BaseViewModel";
import { ConsumableMaterialDefinition, Material, MaterialDocumentation, ProcedureDefinition, MaterialProcedures, StockLocation, RevenueSubAccountCodeDefinition } from "NebulaClient/Model/AtlasClientModel";
import { MaterialSpecialty } from "NebulaClient/Model/AtlasClientModel";
import { MaterialVatRate } from "NebulaClient/Model/AtlasClientModel";
import { MaterialProductLevel } from "NebulaClient/Model/AtlasClientModel";
import { MaterialSpecialtyDef } from "NebulaClient/Model/AtlasClientModel";
import { MaterialPurposeDefinition } from "NebulaClient/Model/AtlasClientModel";
import { MaterialPlaceOfUseDef } from "NebulaClient/Model/AtlasClientModel";
import { SpecialityDefinition } from "NebulaClient/Model/AtlasClientModel";
import { Producer } from "NebulaClient/Model/AtlasClientModel";
import { GMDNDefinition } from "NebulaClient/Model/AtlasClientModel";
import { ProductDefinition } from "NebulaClient/Model/AtlasClientModel";
import { StockCard } from "NebulaClient/Model/AtlasClientModel";
import { DistributionTypeDefinition } from "NebulaClient/Model/AtlasClientModel";
import { MaterialTreeDefinition } from "NebulaClient/Model/AtlasClientModel";
import { Guid } from "app/NebulaClient/Mscorlib/Guid";
import { MalzemeTuru } from "NebulaClient/Model/AtlasClientModel";
import { Type } from "app/NebulaClient/ClassTransformer";
export class ConsumableMaterialDefinitionFormViewModel extends BaseViewModel {
    public _ConsumableMaterialDefinition: ConsumableMaterialDefinition = new ConsumableMaterialDefinition();
    public MaterialSpecialtysGridList: Array<MaterialSpecialty> = new Array<MaterialSpecialty>();
    public MaterialVatRatesGridList: Array<MaterialVatRate> = new Array<MaterialVatRate>();
    public MaterialProductLevelsGridList: Array<MaterialProductLevel> = new Array<MaterialProductLevel>();
    public MaterialSpecialtyDefs: Array<MaterialSpecialtyDef> = new Array<MaterialSpecialtyDef>();
    public MaterialPurposeDefinitions: Array<MaterialPurposeDefinition> = new Array<MaterialPurposeDefinition>();
    public MaterialPlaceOfUseDefs: Array<MaterialPlaceOfUseDef> = new Array<MaterialPlaceOfUseDef>();
    public ConsumableMaterialDefinitions: Array<ConsumableMaterialDefinition> = new Array<ConsumableMaterialDefinition>();
    public SpecialityDefinitions: Array<SpecialityDefinition> = new Array<SpecialityDefinition>();
    public Producers: Array<Producer> = new Array<Producer>();
    public GMDNDefinitions: Array<GMDNDefinition> = new Array<GMDNDefinition>();
    public ProductDefinitions: Array<ProductDefinition> = new Array<ProductDefinition>();
    public StockCards: Array<StockCard> = new Array<StockCard>();
    public DistributionTypeDefinitions: Array<DistributionTypeDefinition> = new Array<DistributionTypeDefinition>();
    public MaterialTreeDefinitions: Array<MaterialTreeDefinition> = new Array<MaterialTreeDefinition>();
    public MalzemeTurus: Array<MalzemeTuru> = new Array<MalzemeTuru>();
    public MaterialPriceDetail: MaterialPriceDetail = new MaterialPriceDetail();
    @Type(() => ProcedureDefinition)
    public ProcedureDefinitions: Array<ProcedureDefinition> = new Array<ProcedureDefinition>();
    @Type(() => MaterialProcedures)
    public grdConsumableMaterialProceduresGridList: Array<MaterialProcedures> = new Array<MaterialProcedures>();

    public MaterialDocumentations: Array<MaterialDocumentation> = new Array<MaterialDocumentation>();
    public MaterialSpecialtysList: Array<Guid> = new Array<Guid>();

    public ConsMaterialEquvalentList: Array<Guid> = new Array<Guid>();
    public MaterialVatRateIn: number;

    @Type(() => Guid)
    public ProdureDefs: Array<Guid> = new Array<Guid>();

    @Type(() => RevenueSubAccountCodeDefinition)
    public RevenueSubAccountCodeDef: RevenueSubAccountCodeDefinition = new RevenueSubAccountCodeDefinition();


    public MaterialTreeDefList: Array<MaterialTreeDefinition> = new Array<MaterialTreeDefinition>();
    public MaterialNewPrice: number;
}

export class ConsumableMaterialDefinitionQueryFilter {
    public IsActive: number = -1;
    public ShowZeroOnDrugOrder: number = -1;
    public IsInheld: number = -1;
    public ShowZeroDistributionInfo: number = -1;
    public ActionStatus: number = -1;
    public StoreID: string;
}

export class MaterialPriceDetail {
    public MaterialPrice: number;
}
export class UploadFile_Input {
    public File: any;
    public FileName: string;
    public Material: any;
    public FileUpdateDate: Date;
    public IsNew: Boolean;
    public ObjectID: Guid;
}

export class MaterialProcedures_Output {
    public ObjectID: Guid;
    public Material: Material_Output;
    ProcedureDefinition: ProcedureDefinition_Output;
}
export class Material_Output {
    public ObjectID: Guid;
    public Name: string;
}
export class ProcedureDefinition_Output {
    public Name: string;
    public ObjectID: Guid;
}

export class FirstInStockUnitePrice {
    public stockActionObjectid: Guid;
    public ObjectDefID: Guid;
    public ExpirationDate: Date;
    public TransactionDate: Date;
    public Tif: string;
    public UnitePrice: string;
}

export class pricePatienList {
    public AccTrxObjID: string;
    public PatientNameAndSurname: string;
    public PatientProtocolNo: string;
    public TrasnactionDate: Date;
    public OldPrice: string;
    public NewPrice: string;
    public Desciption: string;
}

export class pricePatientOutputDTO {
    public unUpdatedPatienList: Array<pricePatienList>;
    public UpdatedPatienList: Array<pricePatienList>;
    public totalUpdatePatientPriceCount: string;
    public priceEndDate: Date;
    public priceStartDate: Date;
}

export class FirstInStockUnitePriceConsumableMaterial {
    public stockActionObjectid: Guid;
    public ObjectDefID: Guid;
    public ExpirationDate: Date;
    public TransactionDate: Date;
    public Tif: string;
    public UnitePrice: string;
}

export class LoadModelComponentConsumableMaterial {
    public shelfAndRowOnStock: ShelfAndRowOnStock;
    public getStockLocationClasses: any[];
    public materialPrices: any[];
    public equivalentMaterialList: any[];
    public getCritical: any;
    public materialProcedures: any[];
    public firstInStockUnitePrices: any[];
    public logDataSources: any[];
    public doSearchaccountingTerm: any;
    public drugSpecifications: any[];
    public MaterialTypeListIsGroupList: any[];
    public MaterialTreeDefinitions: any[];
    public MaterialSpecialtyList: any[];


}

export class ShelfAndRowOnStock {
    public Shelf: number;
    public Row: number;
    public StockLocationID: Guid;
    public StockLocation: StockLocation;
}

export class LoadModelComponentConsumableMaterial_Input {
    public MaterialID: Guid;
    public StoreID: Guid;
    public Shelf: number;
    public Row: number;
}

export class UpdateMaterialPriceOutput {
    public Price: number;
    public Desciption: string;
    public Code: string;
    public SutAppandix: string;
    public ActivationData: string;
}

export class MaterialPriceByMedulaSUT_Input {
    public updateMaterialPriceOutput: UpdateMaterialPriceOutput;
}


export class RepaitUnUpdate_Intput {
    public unUpdatedPatienList: Array<pricePatienList>;
}

/* export class AuditDataSource {
    public lastUpdateDate: Date;
    public lastUpdateUser: string;
    public creationDate: Date;
    public creationUser: string;
} */

File: AtlasWeb/Core/Modules/Merkezi_Yonetim_Sistemi/Stok_Modulleri/Sarf_Edilebilen_Malzeme_Tanimlari/ConsumableMaterialDefinitionFormViewModel.cs
//$7DD0A23D
using System;
using System.Linq;
using Core.Models;
using Infrastructure.Filters;
using Infrastructure.Models;
using TTInstanceManagement;
using TTObjectClasses;
using TTDefinitionManagement;
using TTUtils;
using TTDataDictionary;
using System.Collections.Generic;
using System.ComponentModel;
using Core.Controllers;

namespace Core.Controllers
{
    public partial class ConsumableMaterialDefinitionServiceController
    {
        partial void AfterContextSaveScript_ConsumableMaterialDefinitionForm(ConsumableMaterialDefinitionFormViewModel viewModel, ConsumableMaterialDefinition consumableMaterialDefinition, TTDefinitionManagement.TTObjectStateTransitionDef transDef, TTObjectContext objectContext)
        {
            if (consumableMaterialDefinition.CreationDate == null || ((DateTime)consumableMaterialDefinition.CreationDate).ToShortDateString() == Common.RecTime().ToShortDateString())
            {
                /*if (!(consumableMaterialDefinition.MaterialPrices.Count > 0))
                    updateGetMaterialPriceByIlacAra(consumableMaterialDefinition, objectContext);
                    Sitem direk gidip yapmamamlý ?? 
                */
            }


            objectContext.Save();
        }

        public void updateGetMaterialPriceByIlacAra(ConsumableMaterialDefinition consumableMaterialDefinition, TTObjectContext context)
        {
            try
            {
                MedulaYardimciIslemler.guncelSutSorguGirisDVO guncelSutSorguGirisDVO = new MedulaYardimciIslemler.guncelSutSorguGirisDVO();
                guncelSutSorguGirisDVO.saglikTesisKodu = 11069941;
                guncelSutSorguGirisDVO.tarih = null;

                MedulaYardimciIslemler.guncelSutSorguCevapDVO guncelSutSorguCevapDVO = MedulaYardimciIslemler.WebMethods.guncelSutKodlariSync(Sites.SiteLocalHost, guncelSutSorguGirisDVO);
                if (guncelSutSorguCevapDVO.sonucKodu == "0000")
                {

                    MedulaYardimciIslemler.sutFiyatDVO[] sutKodlariList = guncelSutSorguCevapDVO.sutKodlari;
                    foreach (MedulaYardimciIslemler.sutFiyatDVO item in sutKodlariList)
                    {
                        if (item.turu == 5)
                        {
                            UpdateMaterialPriceOutput updateMaterial = new UpdateMaterialPriceOutput();
                            updateMaterial.Price = item.fiyat;
                            updateMaterial.Desciption = item.adi;
                            updateMaterial.Code = item.sutKodu;
                            updateMaterial.SutAppandix = item.listeAdi;
                            updateMaterial.ActivationDate = item.gecerlilikTarihi;
                            if(consumableMaterialDefinition.Code == updateMaterial.Code)
                            {
                                List<PricingDetailDVO> pricingDetailDefinitions = new List<PricingDetailDVO>();
                                pricingDetailDefinitions = PricingDetailDefinitionForMaterial(context, updateMaterial);
                                //Yapýlacak!!
                                context.Save();
                                break;
                            }
                        }
                    }
                }
                
            }
            catch (Exception ex)
            {

            }
        }
    }
}

namespace Core.Models
{
    public partial class ConsumableMaterialDefinitionFormViewModel : BaseViewModel
    {
        public MaterialPriceDetail MaterialPriceDetail { get; set; }

        public List<MaterialDocumentation> MaterialDocumentations { get; set; }

        public RevenueSubAccountCodeDefinition RevenueSubAccountCodeDef { get; set; }

        public List<MaterialTreeDefinition> MaterialTreeDefList { get; set; }

    }

    public class MaterialPriceDetail
    {
        public Currency MaterialPrice { get; set; }
    }


    public class MaterialProcedures_Output
    {
        public Guid ObjectID { get; set; }
        public Material_Output Material { get; set; }

        public ProcedureDefinition_Output ProcedureDefinition { get; set; }
    }

    public class Material_Output
    {
        public string Name { get; set; }

        public Guid ObjectID { get; set; }
    }

    public class ProcedureDefinition_Output
    {
        public string Name { get; set; }

        public Guid ObjectID { get; set; }

        public string Code { get; set; }
        public string ID { get; set; }
    }

    public class LoadModelComponentConsumableMaterial
    {
        public ShelfAndRowOnStock shelfAndRowOnStock { get; set; }
        public BindingList<StockLocation.GetStockLocation_Class> getStockLocationClasses { get; set; }

        public BindingList<MaterialPrice.MaterialPriceByMaterialForDefinition_Class> materialPrices { get; set; }

        public GetCriticalStockLevels_Class getCritical { get; set; }

        public List<ProcedureDefinition_Output> materialProcedures { get; set; }
        public List<FirstInStockUnitePrice> firstInStockUnitePrices { get; set; }

        public List<LogDataSource> logDataSources { get; set; }

        public LogisticDashboardViewModel doSearchaccountingTerm { get; set; }

        public List<MaterialType_Output> MaterialTypeListIsGroupList { get; set; }
        public List<MaterialTreeDefinition> MaterialTreeDefinitions { get; set; }

        public List<MaterialSpecialty> MaterialSpecialtyList { get; set; }

        public List<EquivalentConsMaterial> equivalentMaterialList { get; set; }
    }

    public class MaterialType_Output
    {
        public string MaterialTypeCode { get; set; }
        public Guid? MaterialTypeObjectID { get; set; }
        public string MaterialTypeName { get; set; }

    }

    public class FirstInStockUnitePrice
    {
        public Guid stockActionObjectid { get; set; }
        public Guid ObjectDefID { get; set; }
        public DateTime? ExpirationDate { get; set; }
        public DateTime? TransactionDate { get; set; }
        public string Tif { get; set; }
        public string UnitePrice { get; set; }
    }

    public class FirstIN_Input
    {
        public Guid MaterialObjectID { get; set; }
        public Guid StoreObjcetID { get; set; }
    }

    public class LoadModelComponentConsumableMaterial_Input
    {
        public Guid MaterialID { get; set; }
        public Guid StoreID { get; set; }
        public int? Shelf { get; set; }
        public int? Row { get; set; }
        public string Equivalent { get; set; }
    }


    public class UpdateMaterialPriceOutput
    {
        public double Price { get; set; }
        public string Desciption { get; set; }
        public string Code { get; set; }
        public string SutAppandix { get; set; }
        public string ActivationDate { get; set; }

    }


    public class MaterialPriceByMedulaSUT_Input
    {
        public UpdateMaterialPriceOutput updateMaterial { get; set; }
        // public ConsumableMaterialDefinition consumableMaterialDefinition { get; set; }
    }

    public class PricingDetailDVO
    {
        public PricingDetailDefinition oldPricingDetailDefinition { get; set; }
        public PricingDetailDefinition newPricingDetailDefinition { get; set; }
    }

    public class pricePatientOutputDTO
    {
        public List<pricePatienList> unUpdatedPatienList { get; set; }
        public List<pricePatienList> UpdatedPatienList { get; set; }
        public string totalUpdatePatientPriceCount { get; set; }
        public DateTime priceEndDate { get; set; }
        public DateTime priceStartDate { get; set; }
    }
    public class RepaitUnUpdate_Intput
    {
        public List<pricePatienList> unUpdatedPatienList { get; set; }
    }
    public class pricePatienList
    {
        public string AccTrxObjID { get; set; }
        public string PatientNameAndSurname { get; set; }
        public string PatientProtocolNo { get; set; }
        public DateTime TrasnactionDate { get; set; }
        public string OldPrice { get; set; }
        public string NewPrice { get; set; }
        public string Desciption { get; set; }
    }

}
